mlflow/pyfunc/backend.py

```diff
@@ -137,33 +274,139 @@ def copy_model_into_container(dockerfile_context_dir):
         )
 
 
-def _execute_in_conda_env(conda_env_path, command, install_mlflow, command_env=None):
+def _execute_in_conda_env(conda_env_path, command, install_mlflow, env_id, command_env=None):
     if command_env is None:
         command_env = os.environ
-    env_id = os.environ.get("MLFLOW_HOME", VERSION) if install_mlflow else None
     conda_env_name = get_or_create_conda_env(conda_env_path, env_id=env_id)
     activate_conda_env = get_conda_command(conda_env_name)
     if install_mlflow:
-        if "MLFLOW_HOME" in os.environ:  # dev version
-            install_mlflow = "pip install -e {} 1>&2".format(os.environ["MLFLOW_HOME"])
-        else:
-            install_mlflow = "pip install mlflow=={} 1>&2".format(VERSION)
+        activate_conda_env += [_get_install_mlflow_command()]
+    _logger.info("=== Running command '%s'", command)
+    _run_multiple_commands(
+        [
+            *activate_conda_env,
+            command,
+        ],
+        env=command_env,
+    )
+
+
+def _validate_pyenv_is_available():
+    # TODO
+    pass
+
+
+def _validate_virtualenv_is_available():
+    # TODO
+    pass
+
+
+def _get_env_name_for_virtualenv(string):
+    return "mlflow-" + hashlib.sha1(string.encode("utf-8")).hexdigest()
+
+
+def _get_mlflow_env_root():
+    return os.getenv(
+        _MLFLOW_ENV_ROOT_ENV_VAR, os.path.join(os.path.expanduser("~"), ".mlflow", "envs")
+    )
+
 
-        activate_conda_env += [install_mlflow]
-    if os.name != "nt":
-        separator = " && "
+def _install_python(version):
+    is_windows = _is_windows()
+    _logger.info("Installing python %s", version)
+    # pyenv-win doesn't support `--skip-existing` but its behavior is enabled by default
+    # https://github.com/pyenv-win/pyenv-win/pull/314
+    pyenv_install_options = () if is_windows else ("--skip-existing",)
+    _run_command(["pyenv", "install", *pyenv_install_options, version])
+
+    if is_windows:
+        # pyenv-win doesn't provide the `pyenv root` command
+        pyenv_root = os.getenv("PYENV_ROOT")
+        if pyenv_root is None:
+            raise MlflowException("Environment variable 'PYENV_ROOT' must be set")
     else:
-        separator = " & "
+        pyenv_root = _get_output(["pyenv", "root"]).strip()
+    return Path(pyenv_root).joinpath("versions", version, "bin", "python")
 
-    command = separator.join(activate_conda_env + [command])
-    _logger.info("=== Running command '%s'", command)
 
-    if os.name != "nt":
-        child = subprocess.Popen(["bash", "-c", command], close_fds=True, env=command_env)
+def _execute_in_virtualenv(local_model_path, command, install_mlflow, env_id, command_env=None):
+    is_windows = _is_windows()
+
+    # Read environment information
+    model_path = Path(local_model_path)
+    python_env_file = model_path / _PYTHON_ENV_FILE_NAME
+    if python_env_file.exists():
+        python_env = PythonEnv.from_yaml(python_env_file)
```

Comment:
Shall we add unit test for `PythonEnv.from_yaml` and `PythonEnv.from_conda_yaml` ?

Reply:
Could you also comment the format of "python_env_file" here ?

Reply:
Added tests and a comment.